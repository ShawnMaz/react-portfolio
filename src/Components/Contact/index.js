import {useState} from 'react';
import {validateEmail} from '../../utils/helper';

function Contact(){
    const [formState, setFormState] = useState({name:'', email:'', message:''});
    const {name, email, message} = formState;

    const [errorMessage, setErrorMessage] = useState('');

    const handleChange = (event) => {
        if(event.target.name === 'email'){
            const isValid = validateEmail(event.target.value);

            if(!isValid){
                setErrorMessage('Invalid email address');
            } else {
                setErrorMessage('');
            }
        } else {
            if(!event.target.value.length){
                setErrorMessage(`${event.target.name} is required`);
            } else {
                setErrorMessage('');
            }
        }

        if(!errorMessage){
            setFormState({...formState, [event.target.name]:event.target.value});
        }

    };

    const handleSubmit = (event) => {
        // prevents the page from refreshing after submitting
        event.preventDefault();
        // resets the form after submitting
        event.target.reset();
        // resets the form state
        setFormState({name:'', email:'', message:''});
    };

    return (
        <section className="container mt-3">
            <h2>Contact</h2>
            <div className="row">
                <form id='contact-form' className="col-sm-12 col-md-6" onSubmit={handleSubmit} noValidate>
                    <div className="mb-3">
                        <label data-testid='nameLabel' htmlFor="name" className="form-label">Name:</label>
                        <input data-testid='nameInput' type='text' name='name' className='form-control' defaultValue={name} placeholder='Full name' onBlur={handleChange}/>
                    </div>
                    <div className="mb-3">
                        <label data-testid='emailLabel' htmlFor="email" className="form-label">Email:</label>
                        <input data-testid='emailInput' type='email' name='email' className='form-control' defaultValue={email} placeholder='Email address' onBlur={handleChange}/>
                    </div>
                    <div className="mb-3">
                        <label data-testid='messageLabel' htmlFor="message" className="form-label">Message:</label>
                        <textarea data-testid='messageTextArea' rows='5' name='message' className='form-control' defaultValue={message} placeholder="Your message" onBlur={handleChange}/>
                    </div>
                    {
                        errorMessage && (
                            <div>
                                <p className='text-danger'>{errorMessage}</p>
                            </div>
                        )
                    }
                    <button data-testid='submitBtn' type='submit' className="btn btn-secondary">Submit</button>
                </form>
            </div>
        </section>
    );
}

export default Contact;