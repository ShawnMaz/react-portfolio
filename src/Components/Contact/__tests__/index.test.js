import {render, cleanup, screen} from '@testing-library/react';
import '@testing-library/jest-dom';
import Contact from '..';

afterEach(cleanup);

describe('Contact component', () => {
    it('renders', () => {
        render(<Contact/>);
    });

    it('matches snapshot', () => {
        const {asFragment} = render(<Contact/>);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Contact component elements are rendered correctly', () => {
    it('renders name, email, message and submit button', () => {
        render(<Contact/>);
        expect(screen.getByTestId('nameLabel')).toHaveTextContent('Name:');
        expect(screen.getByTestId('nameInput')).toBeInTheDocument();
        expect(screen.getByTestId('emailLabel')).toHaveTextContent('Email:');
        expect(screen.getByTestId('emailInput')).toBeInTheDocument();
        expect(screen.getByTestId('messageLabel')).toHaveTextContent('Message:');
        expect(screen.getByTestId('messageTextArea')).toBeInTheDocument();
    });
})