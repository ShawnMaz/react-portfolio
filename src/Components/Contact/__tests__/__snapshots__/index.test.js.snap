// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact component matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="container mt-3"
  >
    <h2>
      Contact
    </h2>
    <div
      class="row"
    >
      <form
        class="col-sm-12 col-md-6"
        id="contact-form"
        novalidate=""
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            data-testid="nameLabel"
            for="name"
          >
            Name:
          </label>
          <input
            class="form-control"
            data-testid="nameInput"
            name="name"
            placeholder="Full name"
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            data-testid="emailLabel"
            for="email"
          >
            Email:
          </label>
          <input
            class="form-control"
            data-testid="emailInput"
            name="email"
            placeholder="Email address"
            type="email"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            data-testid="messageLabel"
            for="message"
          >
            Message:
          </label>
          <textarea
            class="form-control"
            data-testid="messageTextArea"
            name="message"
            placeholder="Your message"
            rows="5"
          />
        </div>
        <button
          class="btn btn-secondary"
          data-testid="submitBtn"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </section>
</DocumentFragment>
`;
